File: nextjs_frontend/lib/api/submissions.ts
/**
 * API Client for Problem Sets and Submissions
 * 
 * This module interacts with the FastAPI backend endpoints for storing
 * and retrieving problem sets and student submissions.
 */

import { get, post, put, del } from './client';

const API_BASE = '/api/py';

// ============================================================================
// PROBLEM SET API
// ============================================================================

/**
 * Get all problem sets from the backend
 */
export async function getAllProblemSets(): Promise<any[]> {
  try {
    const response = await get<{ success: boolean; total: number; problem_sets: any[] }>(
      `${API_BASE}/problem-sets`
    );
    return response.problem_sets || [];
  } catch (error) {
    console.error('Error fetching problem sets:', error);
    return [];
  }
}

/**
 * Get a specific problem set by ID
 */
export async function getProblemSet(problemSetId: string): Promise<any | null> {
  try {
    const response = await get<{ success: boolean; problem_set: any }>(
      `${API_BASE}/problem-sets/${problemSetId}`
    );
    return response.problem_set;
  } catch (error) {
    console.error('Error fetching problem set:', error);
    return null;
  }
}

/**
 * Delete a problem set
 */
export async function deleteProblemSet(problemSetId: string): Promise<void> {
  await del(`${API_BASE}/problem-sets/${problemSetId}`);
}

// ============================================================================
// SUBMISSION API
// ============================================================================

/**
 * Store a student submission
 * Images are now URL references (/api/py/images/IMAGE_ID), not base64
 */
export async function storeSubmission(
  problemSetId: string,
  problemId: number,
  studentName: string,
  solution: string
): Promise<any> {
  const response = await post<{ success: boolean; submission_id: string; submission: any }>(
    `${API_BASE}/submissions`,
    {
      problem_set_id: problemSetId,
      problem_id: problemId,
      student_name: studentName,
      solution: solution  // Contains image URLs like ![](/ api/py/images/img_123)
    }
  );
  return response.submission;
}

/**
 * Get all submissions for a problem set
 */
export async function getSubmissions(
  problemSetId: string,
  problemId?: number
): Promise<any[]> {
  try {
    const url = problemId
      ? `${API_BASE}/submissions/${problemSetId}?problem_id=${problemId}`
      : `${API_BASE}/submissions/${problemSetId}`;

    const response = await get<{ success: boolean; submissions: any[] }>(url);
    return response.submissions || [];
  } catch (error) {
    console.error('Error fetching submissions:', error);
    return [];
  }
}

/**
 * Get a student's submission for a specific problem
 */
export async function getStudentSubmission(
  problemSetId: string,
  problemId: number,
  studentName: string
): Promise<any | null> {
  try {
    const response = await get<{ success: boolean; submissions: any[] }>(
      `${API_BASE}/submissions/${problemSetId}/${studentName}`
    );
    const submission = response.submissions.find((s: any) => s.problem_id === problemId);
    return submission || null;
  } catch (error) {
    console.error('Error fetching student submission:', error);
    return null;
  }
}

/**
 * Update a submission with grade
 */
export async function updateSubmissionGrade(
  problemSetId: string,
  problemId: number,
  studentName: string,
  grade: any
): Promise<void> {
  // First, get the submission ID
  const submissions = await getSubmissions(problemSetId, problemId);
  const submission = submissions.find(s => s.student_name === studentName);

  if (submission) {
    await put(
      `${API_BASE}/submissions/${submission.id}/grade`,
      grade
    );
  }
}

/**
 * Grade all submissions for a specific problem
 */
export async function gradeBatchForProblem(
  problemSetId: string,
  problemId: number
): Promise<{ success: boolean; statistics: any; results: any[] }> {
  return await post<{ success: boolean; statistics: any; results: any[] }>(
    `${API_BASE}/submissions/grade-batch?problem_set_id=${problemSetId}&problem_id=${problemId}`,
    {}
  );
}

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

/**
 * Get all submissions for a problem set (all problems)
 */
export async function getAllSubmissionsForSet(problemSetId: string): Promise<any[]> {
  return await getSubmissions(problemSetId);
}

/**
 * Count submissions for a problem set
 */
export async function countSubmissions(problemSetId: string): Promise<{
  total: number;
  graded: number;
  byProblem: { [problemId: number]: { total: number; graded: number } };
}> {
  const allSubs = await getAllSubmissionsForSet(problemSetId);
  const graded = allSubs.filter(s => s.graded).length;

  // Count by problem
  const byProblem: { [problemId: number]: { total: number; graded: number } } = {};
  allSubs.forEach(sub => {
    const probId = sub.problem_id;
    if (!byProblem[probId]) {
      byProblem[probId] = { total: 0, graded: 0 };
    }
    byProblem[probId].total++;
    if (sub.graded) {
      byProblem[probId].graded++;
    }
  });

  return {
    total: allSubs.length,
    graded,
    byProblem,
  };
}

File: fastapi_backend/api/routers/submissions.py
"""
Submissions router - handles student submissions and grading.
"""
from typing import Any, Dict, List, Optional
import json
import traceback
from datetime import datetime
import uuid

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from api.dependencies import (
    PROBLEM_SETS_DIR,
    SUBMISSIONS_DIR,
    get_grading_orchestrator,
)

router = APIRouter(prefix="/api/py", tags=["submissions"])


# Request models
class StudentSubmission(BaseModel):
    name: str
    solution: str


class GradeSubmissionsRequest(BaseModel):
    problem: Dict[str, Any]
    correct_solution: str
    student_submissions: List[StudentSubmission]


class StoreSubmissionRequest(BaseModel):
    problem_set_id: str
    problem_id: int
    student_name: str
    solution: str
    images: Optional[List[str]] = []  # List of base64 encoded images


@router.post("/grade-submissions")
def grade_submissions(payload: GradeSubmissionsRequest) -> Dict[str, Any]:
    """
    Grade multiple student submissions for a single problem.

    Expects the problem definition and correct solution (for example,
    taken directly from the problem set generation response) as well as
    a list of {name, solution} submissions.
    """
    grader = get_grading_orchestrator()

    # Convert Pydantic models to simple dicts expected by GradingOrchestrator
    student_submissions = [
        {"name": s.name, "solution": s.solution} for s in payload.student_submissions
    ]

    results, stats = grader.grade_batch(
        problem=payload.problem,
        correct_solution=payload.correct_solution,
        student_submissions=student_submissions,
    )

    return {
        "success": True,
        "statistics": stats,
        "results": results,
    }


@router.post("/submissions")
def store_submission(payload: StoreSubmissionRequest) -> Dict[str, Any]:
    """
    Store a student submission.
    
    Args:
        payload: Submission data
        
    Returns:
        Stored submission with ID
    """
    try:
        # Load or create submissions file for this problem set
        submissions_file = SUBMISSIONS_DIR / f"{payload.problem_set_id}.json"
        
        if submissions_file.exists():
            with open(submissions_file, 'r', encoding='utf-8') as f:
                submissions_data = json.load(f)
        else:
            submissions_data = {"problem_set_id": payload.problem_set_id, "submissions": []}
        
        # Create submission
        submission_id = f"sub_{uuid.uuid4().hex[:12]}"
        submission = {
            "id": submission_id,
            "problem_id": payload.problem_id,
            "student_name": payload.student_name,
            "solution": payload.solution,
            "images": payload.images or [],  # Store images separately
            "submitted_at": datetime.now().isoformat(),
            "graded": False,
            "grade": None
        }
        
        # Check if student already submitted this problem
        existing_submissions = submissions_data.get("submissions", [])
        existing_index = None
        for i, sub in enumerate(existing_submissions):
            if (sub.get("student_name") == payload.student_name and 
                sub.get("problem_id") == payload.problem_id):
                existing_index = i
                break
        
        if existing_index is not None:
            # Update existing submission
            submission["id"] = existing_submissions[existing_index].get("id", submission_id)
            existing_submissions[existing_index] = submission
        else:
            # Add new submission
            existing_submissions.append(submission)
        
        submissions_data["submissions"] = existing_submissions
        
        # Save to file
        with open(submissions_file, 'w', encoding='utf-8') as f:
            json.dump(submissions_data, f, indent=2, ensure_ascii=False)
        
        return {
            "success": True,
            "submission_id": submission["id"],
            "submission": submission
        }
    except Exception as e:
        print(f"[API] Error storing submission: {repr(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/submissions/{problem_set_id}")
def get_submissions(problem_set_id: str, problem_id: Optional[int] = None) -> Dict[str, Any]:
    """
    Get all submissions for a problem set.
    
    Args:
        problem_set_id: Problem set ID
        problem_id: Optional problem ID to filter by
        
    Returns:
        List of submissions
    """
    try:
        submissions_file = SUBMISSIONS_DIR / f"{problem_set_id}.json"
        
        if not submissions_file.exists():
            return {
                "success": True,
                "problem_set_id": problem_set_id,
                "submissions": []
            }
        
        with open(submissions_file, 'r', encoding='utf-8') as f:
            submissions_data = json.load(f)
        
        submissions = submissions_data.get("submissions", [])
        
        # Filter by problem_id if provided
        if problem_id is not None:
            submissions = [s for s in submissions if s.get("problem_id") == problem_id]
        
        return {
            "success": True,
            "problem_set_id": problem_set_id,
            "total": len(submissions),
            "submissions": submissions
        }
    except Exception as e:
        print(f"[API] Error getting submissions: {repr(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/submissions/{problem_set_id}/{student_name}")
def get_student_submissions(problem_set_id: str, student_name: str) -> Dict[str, Any]:
    """
    Get a student's submissions for a problem set.
    
    Args:
        problem_set_id: Problem set ID
        student_name: Student name
        
    Returns:
        List of student's submissions
    """
    try:
        submissions_file = SUBMISSIONS_DIR / f"{problem_set_id}.json"
        
        if not submissions_file.exists():
            return {
                "success": True,
                "problem_set_id": problem_set_id,
                "student_name": student_name,
                "submissions": []
            }
        
        with open(submissions_file, 'r', encoding='utf-8') as f:
            submissions_data = json.load(f)
        
        submissions = submissions_data.get("submissions", [])
        student_submissions = [s for s in submissions if s.get("student_name") == student_name]
        
        return {
            "success": True,
            "problem_set_id": problem_set_id,
            "student_name": student_name,
            "submissions": student_submissions
        }
    except Exception as e:
        print(f"[API] Error getting student submissions: {repr(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/submissions/{submission_id}/grade")
def update_submission_grade(submission_id: str, grade: Dict[str, Any]) -> Dict[str, Any]:
    """
    Update a submission with grading results.
    
    Args:
        submission_id: Submission ID
        grade: Grade data from grading API
        
    Returns:
        Updated submission
    """
    try:
        # Find the submission across all problem sets
        for submissions_file in SUBMISSIONS_DIR.glob("*.json"):
            with open(submissions_file, 'r', encoding='utf-8') as f:
                submissions_data = json.load(f)
            
            submissions = submissions_data.get("submissions", [])
            updated = False
            
            for i, sub in enumerate(submissions):
                if sub.get("id") == submission_id:
                    submissions[i]["graded"] = True
                    submissions[i]["grade"] = grade
                    updated = True
                    break
            
            if updated:
                # Save back to file
                with open(submissions_file, 'w', encoding='utf-8') as f:
                    json.dump(submissions_data, f, indent=2, ensure_ascii=False)
                
                return {
                    "success": True,
                    "submission_id": submission_id,
                    "submission": submissions[i]
                }
        
        raise HTTPException(status_code=404, detail=f"Submission '{submission_id}' not found")
    except HTTPException:
        raise
    except Exception as e:
        print(f"[API] Error updating submission grade: {repr(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/submissions/grade-batch")
def grade_batch_and_store(
    problem_set_id: str,
    problem_id: int
) -> Dict[str, Any]:
    """
    Grade all submissions for a specific problem and store results.
    
    This is a convenience endpoint that:
    1. Gets the problem set and problem
    2. Gets all submissions for that problem
    3. Grades them using the grading API
    4. Stores the grades back
    
    Args:
        problem_set_id: Problem set ID
        problem_id: Problem ID to grade
        
    Returns:
        Grading results with statistics
    """
    try:
        # Get problem set
        problem_set_file = PROBLEM_SETS_DIR / f"{problem_set_id}.json"
        if not problem_set_file.exists():
            raise HTTPException(status_code=404, detail=f"Problem set '{problem_set_id}' not found")
        
        with open(problem_set_file, 'r', encoding='utf-8') as f:
            problem_set_data = json.load(f)
        
        # Find the problem
        problems = problem_set_data.get("problem_set", [])
        problem_data = None
        for item in problems:
            if item.get("problem", {}).get("id") == problem_id:
                problem_data = item
                break
        
        if not problem_data:
            raise HTTPException(status_code=404, detail=f"Problem {problem_id} not found in problem set")
        
        problem = problem_data["problem"]
        correct_solution = problem_data["solution"]
        
        # Get submissions
        submissions_file = SUBMISSIONS_DIR / f"{problem_set_id}.json"
        if not submissions_file.exists():
            raise HTTPException(status_code=404, detail="No submissions found")
        
        with open(submissions_file, 'r', encoding='utf-8') as f:
            submissions_data = json.load(f)
        
        problem_submissions = [
            s for s in submissions_data.get("submissions", [])
            if s.get("problem_id") == problem_id
        ]
        
        if not problem_submissions:
            raise HTTPException(status_code=404, detail="No submissions for this problem")
        
        # Format for grading
        student_submissions = [
            {
                "name": s["student_name"], 
                "solution": s["solution"],
                "images": s.get("images", [])  # Include images for vision-based grading
            }
            for s in problem_submissions
        ]
        
        # Grade them
        grader = get_grading_orchestrator()
        results, stats = grader.grade_batch(
            problem=problem,
            correct_solution=correct_solution,
            student_submissions=student_submissions
        )
        
        # Store grades back
        all_submissions = submissions_data.get("submissions", [])
        for result in results:
            student_name = result["student_name"]
            for i, sub in enumerate(all_submissions):
                if (sub.get("student_name") == student_name and 
                    sub.get("problem_id") == problem_id):
                    all_submissions[i]["graded"] = True
                    all_submissions[i]["grade"] = result
                    break
        
        submissions_data["submissions"] = all_submissions
        
        # Save back
        with open(submissions_file, 'w', encoding='utf-8') as f:
            json.dump(submissions_data, f, indent=2, ensure_ascii=False)
        
        return {
            "success": True,
            "statistics": stats,
            "results": results
        }
    except HTTPException:
        raise
    except Exception as e:
        print(f"[API] Error in grade_batch_and_store: {repr(e)}")
        print(f"[API] Traceback:\n{traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=str(e))
